vllm/spec_decode/multi_step_worker.py

```diff
@@ -24,14 +29,67 @@ class MultiStepWorker(Worker, ProposerWorkerBase):
     requires more thought for MultiStepWorker support.
     """
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
+    def __init__(self, draft_ranks: Optional[List[int]] = None, **kwargs):
+        """Create a MultiStepWorker.
+
+        It allows a speculative draft model to run with smaller tensor
+        parallel degree than target model.
+        This reduces the communication overhead of small draft models.
+
+        This is implemented by changing vLLM's tensor parallel group to a group
+        of the small size temporarily during forward passes of draft models.
+
+        Args:
+            draft_ranks (Optional[List[int]]): if this value is given, only some
+            of the GPU ranks in this value participate in draft generation
+        """
+
+        self._draft_ranks = draft_ranks
+        self._is_dummy = False
+        self._world_group = None
+        self._tp_group = None
+
+        if draft_ranks is not None:
+            # whether the worker participates in draft generation or not
+            self._is_dummy = kwargs['rank'] not in draft_ranks
+
+        super().__init__(**kwargs)
 
         # Lazy initialization list.
-        self._proposer: Top1Proposer
+        self._proposer: SpeculativeProposer
+
+    def _patch_tensor_parallel_group(self):
+        return patch_tensor_parallel_group(self._world_group, self._tp_group)
 
     def init_device(self):
-        super().init_device()
+        if self._is_dummy:
+            return
+
+        if self._draft_ranks is not None:
+            # creates tp process group containing only a subset of gpu ranks
+            local_rank = get_world_group().local_rank
+            world_backend = torch.distributed.get_backend(
+                get_world_group().device_group)
+            tp_backend = torch.distributed.get_backend(
+                get_tp_group().device_group)
+
+            self._world_group = GroupCoordinator(
+                group_ranks=[self._draft_ranks],
+                local_rank=local_rank,
+                torch_distributed_backend=world_backend,
+                use_pynccl=False,
+                use_custom_allreduce=False,
+            )
+            self._tp_group = GroupCoordinator(
+                group_ranks=[self._draft_ranks],
+                local_rank=local_rank,
+                torch_distributed_backend=tp_backend,
+                use_pynccl=True,
+                use_custom_allreduce=_ENABLE_CUSTOM_ALL_REDUCE,
+            )
```

Comment:
Can we move the construction of these objects to parallel_state? We shouldn't import private variables from the package (`_ENABLE_CUSTOM_ALL_REDUCE`)

Also want to keep magic settings like `use_pynccl` close together / same file

Reply:
I've moved them into paralell_state.py and added init methods for world and tp/pp groups.